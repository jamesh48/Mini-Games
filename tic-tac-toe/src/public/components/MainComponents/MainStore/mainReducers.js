const reduceReducers = (...reducers) => {
  return (state, action) => reducers.reduce((acc, nextReducer) => nextReducer(acc, action), state)
};

const hasAlreadyWon = (state = null, action) => {
  switch (action.type) {
    case 'TOGGLE HAS ALREADY WON TRUE':
      return action.payload;
    default:
      return state;
  };
};


const xPlayerName = (state = null, action) => {
  switch (action.type) {
    case 'SUBMIT X PLAYER NAME':
      return action.payload;
    default:
      return state;
  };
};

const oPlayerName = (state = null, action) => {
  switch (action.type) {
    case 'SUBMIT O PLAYER NAME':
      return action.payload;
    default:
      return state;
  }
}

const combineReducers = (slices) => (state, action) => Object.keys(slices).reduce((acc, prop) => ({
  ...acc,
  [prop]: slices[prop](acc[prop], action),
}),
  state
);

export default combineReducers({ hasAlreadyWon, xPlayerName, oPlayerName });

