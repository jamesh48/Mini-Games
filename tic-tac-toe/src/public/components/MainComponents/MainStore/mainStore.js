// https://stackoverflow.com/questions/59200785/react-usereducer-how-to-combine-multiple-reducers
import React, { useReducer } from 'react';
import CombinedReducers from './mainReducers.js';
const MainStoreContext = React.createContext();

const initialState = {
  hasAlreadyWon: null,
  xPlayerName: null,
  oPlayerName: null
};

const MainStoreProvider = ({ children }) => {
  const [state, dispatch] = React.useReducer(CombinedReducers, initialState);
  const store = React.useMemo(() => [state, dispatch], [state]);
  return (
    <MainStoreContext.Provider value={store}>{children}</MainStoreContext.Provider>
  );
};

const useMainContext = () => {
  const context = React.useContext(MainStoreContext);
  if (context === undefined) {
    throw new Error('useMainStoreContext must be used within a MainStoreProvider')
  }
  return context;
};


export { useMainContext, MainStoreProvider };