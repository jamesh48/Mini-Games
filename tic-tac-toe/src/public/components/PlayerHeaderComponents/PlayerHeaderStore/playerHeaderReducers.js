const reduceReducers = (...reducers) => {
  return (state, action) => reducers.reduce((acc, nextReducer) => nextReducer(acc, action), state)
};

const xPlayerWinCount = (state = 0, action) => {
  switch (action.type) {
    case 'INCREMENT X PLAYER WIN COUNT':
      return state + 1;
    default:
      return state;
  };
};

const oPlayerWinCount = (state = 0, action) => {
  switch (action.type) {
    case 'INCREMENT O PLAYER WIN COUNT':
      return state + 1;
    default:
      return state;
  };
};

const combineReducers = (slices) => (state, action) => Object.keys(slices).reduce((acc, prop) => ({
  ...acc,
  [prop]: slices[prop](acc[prop], action),
}),
  state
);

export default combineReducers({ xPlayerWinCount, oPlayerWinCount });

