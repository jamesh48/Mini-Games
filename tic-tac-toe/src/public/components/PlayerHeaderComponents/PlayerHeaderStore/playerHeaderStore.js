// https://stackoverflow.com/questions/59200785/react-usereducer-how-to-combine-multiple-reducers
import React, { useReducer } from 'react';
import CombinedReducers from './playerHeaderReducers.js';

const PlayerHeaderContext = React.createContext();

const initialState = {
  oPlayerWinCount: 0,
  xPlayerWinCount: 0
};

const PlayerHeaderStoreProvider = ({ children }) => {
  const [state, dispatch] = React.useReducer(CombinedReducers, initialState);
  const store = React.useMemo(() => [state, dispatch], [state]);
  return (
    <PlayerHeaderContext.Provider value={store}>{children}</PlayerHeaderContext.Provider>
  );
};

const usePlayerHeaderContext = () => {
  const context = React.useContext(PlayerHeaderContext);
  if (context === undefined) {
    throw new Error('usePlayerHeaderContext must be used within a PlayerHeaderStoreProvider');
  }
  return context;
};


export { usePlayerHeaderContext, PlayerHeaderStoreProvider };