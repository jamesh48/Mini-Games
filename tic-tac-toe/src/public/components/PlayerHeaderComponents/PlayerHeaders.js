import React from 'react';
import { useMainContext } from 'MainStore';
import { usePlayerHeaderContext } from 'PlayerHeaderStore';

import './playerHeaders.scss';

export default () => {
  const [{ xPlayerName, oPlayerName }] = useMainContext();
  const [{ xPlayerWinCount, oPlayerWinCount }] = usePlayerHeaderContext();

  return (
    <div id='player-headers'>
      <span className='player-name-headers' id='p1-name-header'>{xPlayerName || 'X Player-> '}
        <h1 className='player-name-tally' id='p-x-tally'>{xPlayerWinCount}</h1>
      </span>
      <span className='player-name-headers' id='p2-name-header'>{oPlayerName || 'O Player-> '}
        <p className='player-name-tally' id='p-o-tally'>{oPlayerWinCount}</p>
      </span>
    </div>
  );
};