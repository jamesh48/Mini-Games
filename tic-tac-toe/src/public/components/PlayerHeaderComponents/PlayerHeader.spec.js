import React from 'react';
import { MainStoreProvider } from 'MainStore';
import PlayerHeaders from './PlayerHeaders.js';
import { PlayerHeaderStoreProvider } from 'PlayerHeaderStore';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

describe('Player Headers', () => {

  it('renders the player header component and its children', () => {
    const wrapper = mount(
      <MainStoreProvider>
        <PlayerHeaderStoreProvider>
          <PlayerHeaders />
        </PlayerHeaderStoreProvider>
      </MainStoreProvider>);
    expect(wrapper.find('#player-headers').length).toEqual(1)
    expect(wrapper.find('.player-name-headers').length).toEqual(2);
  });

  it('initializes player win count at 0 each', () => {
    const wrapper = mount(
      <MainStoreProvider>
        <PlayerHeaderStoreProvider>
          <PlayerHeaders />
        </PlayerHeaderStoreProvider>
      </MainStoreProvider>
    );
    expect(wrapper.text()).toContain('X Player-> 0');
    expect(wrapper.text()).toContain('O Player-> 0');
  });

  test('snapshot renders', () => {
    const component = renderer.create(
      <MainStoreProvider>
        <PlayerHeaderStoreProvider>
          <PlayerHeaders />
        </PlayerHeaderStoreProvider>
      </MainStoreProvider >
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});

