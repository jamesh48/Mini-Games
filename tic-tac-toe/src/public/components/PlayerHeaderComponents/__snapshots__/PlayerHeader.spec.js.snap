// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player Headers snapshot renders 1`] = `
<div
  id="player-headers"
>
  <span
    className="player-name-headers"
    id="p1-name-header"
  >
    X Player-&gt; 
    <h1
      className="player-name-tally"
      id="p-x-tally"
    >
      0
    </h1>
  </span>
  <span
    className="player-name-headers"
    id="p2-name-header"
  >
    O Player-&gt; 
    <p
      className="player-name-tally"
      id="p-o-tally"
    >
      0
    </p>
  </span>
</div>
`;
